Adds blog post for pluggable storage (#14003)
supabase/supabase: apps/www/_blog/2023-04-27-postgres-pluggable-strorage.mdx

---

> title: 'Postgres pluggable storage'

Reviewer: When I see the title, I think the blog post will be a highly technical one but this is mostly about history.

How about, `"History of Postgres pluggable storage"`?

description: Exploring history of Postgres pluggable storage and the possibility of landing it in the Postgres core.
author: paul_copplestone
image: lw7-hackathon-winners/lw7-hackathon-winner.jpg
thumb: lw7-hackathon-winners/lw7-hackathon-winner.jpg
tags:
  - database
date: '2023-04-27'
toc_depth: 2
---

You might have recently read “[The Part of PostgreSQL We Hate the Most](https://ottertune.com/blog/the-part-of-postgresql-we-hate-the-most/)” which highlights some of the deficiencies with Postgres storage. Luckily, there is some work in Postgres ecosystem to solve this using “pluggable storage”.

This article explores the history of Postgres pluggable storage and the possibility of landing it in the Postgres core.

## What is pluggable storage?

Pluggable storage gives developers the ability to use different storage engines for different tables within the same database.

> Developers will be able to choose a storage method that is optimized for their specific needs. Some tables could be optimized for high transactional loads, others for analytics workloads, while yet another for archiving.

Reviewer 2: ```suggestion
Developers will be able to choose a storage method that is optimized for their specific needs. Some tables could be configured for high transactional loads, others for analytics workloads, and still others for archiving.
```

Something like this is already [available in MySQL](https://en.wikipedia.org/wiki/Comparison_of_MySQL_database_engines), which uses the [InnoDB](https://en.wikipedia.org/wiki/InnoDB) as the default storage engine since MySQL 5.5. This replaced [MyISAM](https://en.wikipedia.org/wiki/MyISAM) which had it’s own [set of problems](https://www.percona.com/blog/using-myisam-in-production/).

The “pluggable” part refers to the ability for developers to develop their own storage methods, similar to Postgres Extensions. While some databases offer multiple built-in storage methods by default, Postgres follows a core philosophy of flexibility and extensibility, and using a pluggable approach makes a lot of sense.

## Current progress for pluggable storage

> PostgreSQL received basic support for pluggable storage in version 12 with the goal of [adding ZHeap](https://anarazel.de/talks/2019-05-30-pgcon-pluggable-table-storage/pluggable.pdf). Andres Freund and several contributors provided some excellent refactoring, improving various APIs for table access methods (TAM).

Reviewer 2: ```suggestion
In version 12, PostgreSQL introduced basic support for pluggable storage with the goal of [adding ZHeap](https://anarazel.de/talks/2019-05-30-pgcon-pluggable-table-storage/pluggable.pdf). Andres Freund and several contributors enabled that feature by refactoring and expanding a variety of table access method (TAM) APIs.
```

While It was a great start, it appears that the [work on ZHeap has stalled](https://twitter.com/andy_pavlo/status/1590703943176589312).

Even without a full API for pluggable storage, there are a couple of solutions available on the market today:

- **Citus’s** [Columnar Storage](https://docs.citusdata.com/en/v11.1/admin_guide/table_management.html#columnar-storage), which powers Microsoft’s [Hyperscale](https://azure.microsoft.com/en-us/updates/azure-database-for-postgresql-hyperscale-citus-is-now-available/) offering. Citus has a few [limitations](https://docs.citusdata.com/en/v11.1/admin_guide/table_management.html#limitations) due to the limitations of the current API.
- EDB’s [Advanced Storage Pack](https://www.infoworld.com/article/3681112/3-key-features-in-edb-postgresql-15.html), which is appears to provide some enhancements to heap storage, although the source is not available for us to confirm.

## The future of pluggable storage

Right now, pluggable storage requires a a fork of Postgres. But there are several initiatives in the Postgres community that we’re excited about.

> - There have been community updates for [custom resource managers](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=5c279a6d350205cc98f91fb8e1d3e4442a6b25d1) that expand WAL support for TAM.

Reviewer 2: ```suggestion
- There have been community updates for [custom resource managers](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=5c279a6d350205cc98f91fb8e1d3e4442a6b25d1) that expand [write-ahead log (WAL)](https://www.postgresql.org/docs/current/wal-intro.html) support for TAM.
```

> - The creators at [OrioleDB](https://github.com/orioledb/orioledb) started with a fork of Postgres and have been quietly upstreaming the code that is required to make pluggable storage a reality. They are solving many “[wicked Postgres problems](https://www.slideshare.net/AlexanderKorotkov/solving-postgresql-wicked-problems)” with [dual pointers](https://www.orioledata.com/blog/buffer-management/) instead of buffer mapping, row-level WAL instead of block-level WAL, undo logs instead of bloat-prone MVCC, and index-organized tables instead of heap. In Postgres 14 their patchset was about 5,000 lines of code (this represents the total number of changes required in their fork). In the upcoming Postgres 16 release, it is around 2,000 lines of code. That means that around 60% of the code is already committed to the Postgres core. (Disclosure: Supabase started [sponsoring OrioleDB](https://supabase.com/blog/supabase-series-b#where-were-going) in 2022).

Reviewer 2: Lots of content here. Possibly break this out into a paragraph vs a bullet point

<aside>
ℹ️ Alexander Korotkov, the maintainer of OrioleDB, will be talking at PGCON 23 about some of the [changes](https://www.pgcon.org/events/pgcon_2023/schedule/session/470-future-of-table-access-methods/) they are making.

</aside>

## Timelines for pluggable storage

Ambitiously, many of the changes required for pluggable storage could land in Postgres 17. If that happens, new storage engines like OrioleDB would be available as a simple extension for developers to use in any Postgres instance. This paves the way for other Pluggable Storage Engines, just like MySQL’s [support for multiple storage engines](https://dev.mysql.com/doc/refman/5.7/en/storage-engines.html).

> As a community project, Postgres mostly requires support for these initiatives. The best way to support these initiatives is to get involved with Postgres [Commitfests](https://supabase.com/blog/postgresql-commitfest). Postgres is already one of the best database engines in the world, and pluggable storage would make it even more attractive and maintanable.

Author: ```suggestion
As a community project, Postgres requires support for these initiatives. You can support pluggable storage by getting involved during Postgres [Commitfests](https://supabase.com/blog/postgresql-commitfest). Postgres is already one of the best database engines in the world, and pluggable storage would make it even more attractive and maintainable.
```
